### PR TITLE
Add side staking cap tests

### LINKED ISSUE #753: Do we thoroughly test when SideStaking bot goes past cap? If no, test
**Is your feature request related to a problem? Please describe.**
SideStaking bot (ssbot) has a datatoken cap. Up to that cap, for whatever OCEAN comes in, it can release datatokens such that price stays constant (ie the ratio of datatoken:OCEAN is constant). After it hits the cap, it no longer matches with datatokens.

Developers using ocean.py should be able to know what behavior to expect. 

And, ocean.py should behave as expected, for happy path & corner cases.

Example questions:
- after it hits the cap, then enough OCEAN is removed from pool to go below cap, then does it go back to matching 1:1, or not?
- after it hits the cap, there should be much higher slippage with more OCEAN liquidity going in, and OCEAN liquidity going out
- are there opportunities for malicious actors to hurt others? Eg publisher rug pulls? Show tests of potential attacks. (BTW I believe we should be ok. But the test should show everyone's ok under attempted rug pulls.)

**Describe the solution you'd like**

- Test happy path
- Test corner cases

### PR DESCRIPTION
**Pull request by md0x**
Signed-off-by: Pablo Maldonado <pablo@oceanprotocol.com>

Fixes #753.

Changes proposed in this PR:

- Test the cap limits of the Side Staking bot
- Try to steal base token from the pool unsuccessful
- Check the "constant" ratio-> To be discussed, see my TODO

### CODE REVIEW

File: ocean_lib/models/test/test_side_staking.py [hunk 1]
```diff
@@ -405,3 +414,224 @@ def test_side_staking(
...
+        ocean_token.address, erc20_token.address, swap_market_fee
+    )
+    join_pool_one_side(web3, bpool, ocean_token, another_consumer_wallet, to_wei("0.5"))
+    final_spot_price_basetoken_datatoken = bpool.get_spot_price(
+        ocean_token.address, erc20_token.address, swap_market_fee
+    )
+
+    # TODO: Is this assertion normal?
+    assert (
+        final_spot_price_basetoken_datatoken != initial_spot_price_basetoken_datatoken
```
Comment (md0x) [inline ocean_lib/models/test/test_side_staking.py:548]: The spot price is not identical, there is a 7% change.
Test case: 0 vesting amount, added base token 1 side liquidity when the liquidity was considerably below the cap

File: ocean_lib/models/test/test_side_staking.py [hunk 2]
```diff
@@ -405,3 +413,235 @@ def test_side_staking(
         err.value.args[0]
         == "execution reverted: VM Exception while processing transaction: revert ERR: Only pool can call this"
     )
+
+
+@pytest.mark.unit
+def test_side_staking_steal(
```
Comment (md0x) [inline ocean_lib/models/test/test_side_staking.py:419]: ⚠️⚠️⚠️

File: ocean_lib/models/test/test_side_staking.py [hunk 3]
```diff
@@ -405,3 +413,235 @@ def test_side_staking(
...
+    # End test initial values and setups
+
+    initial_ocean_user_balance = ocean_token.balanceOf(another_consumer_wallet.address)
+    swap_exact_amount_in_base_token(
+        bpool, erc20_token, ocean_token, another_consumer_wallet, to_wei("0.1")
+    )
+
+    # Fill the pool until the cap is reached
+    previous_pool_datatoken_balance = 0
+    while previous_pool_datatoken_balance != erc20_token.balanceOf(bpool.address):
```
Comment (md0x) [inline ocean_lib/models/test/test_side_staking.py:459]: Every `join_pool_one_side` with current setup increases the spot price ~6-7%

File: ocean_lib/models/test/test_side_staking.py [hunk 4]
```diff
@@ -405,3 +413,235 @@ def test_side_staking(
...
+        previous_pool_datatoken_balance = erc20_token.balanceOf(bpool.address)
+        join_pool_one_side(
+            web3, bpool, ocean_token, another_consumer_wallet, to_wei("0.5")
+        )
+    join_pool_one_side(web3, bpool, ocean_token, another_consumer_wallet, to_wei("0.5"))
+
+    swap_exact_amount_in_datatoken(
```
Comment (md0x) [inline ocean_lib/models/test/test_side_staking.py:466]: Here we sell after manipulating the price

File: ocean_lib/models/test/test_side_staking.py [hunk 5]
```diff
@@ -405,3 +413,235 @@ def test_side_staking(
...
+        bpool,
+        erc20_token,
+        ocean_token,
+        another_consumer_wallet,
+        erc20_token.balanceOf(another_consumer_wallet.address),
+    )
+
+    while pool_token.balanceOf(another_consumer_wallet.address) > to_wei("1"):
+        wallet_exit_pool_one_side(
```
Comment (md0x) [inline ocean_lib/models/test/test_side_staking.py:475]: Exiting with one side keeps our profit

File: ocean_lib/models/test/test_side_staking.py [hunk 6]
```diff
@@ -392,3 +387,187 @@ def test_side_staking(
...
+    _, erc20 = deploy_erc721_erc20(
+        web3=web3,
+        config=config,
+        erc721_publisher=publisher_wallet,
+        erc20_minter=publisher_wallet,
+        cap=to_wei(to_wei("1000000000")),
+    )
+
+    erc20.mint(
+        publisher_wallet.address, to_wei("10000000000000000000"), publisher_wallet
```
Comment (DMats) [inline ocean_lib/models/test/test_side_staking.py:423]: Consider using `MAX_WEI`.
https://github.com/oceanprotocol/ocean.py/blob/v4main/ocean_lib/web3_internal/currency.py#L31

File: ocean_lib/models/test/test_side_staking.py [hunk 7]
```diff
@@ -392,3 +387,187 @@ def test_side_staking(
...
+    token_cap_eth = 1  # doesnt matter
+    join_pool_step_eth = 20000
+    datatoken_buy_amount_basetoken_eth = 5000
+    # ---------------------------------------------------------------------------------------------
+
+    initial_pool_liquidity = to_wei(initial_pool_liquidity_eth)
+    token_cap = to_wei(token_cap_eth)
+    swap_market_fee = to_wei("0.0001")
+    swap_fee = to_wei("0.0001")
+    big_allowance = to_wei("10000000000000000")
```
Comment (DMats) [inline ocean_lib/models/test/test_side_staking.py:412]: Consider using `MAX_WEI`
https://github.com/oceanprotocol/ocean.py/blob/v4main/ocean_lib/web3_internal/currency.py#L31

File: ocean_lib/models/test/test_side_staking.py [hunk 8]
```diff
@@ -392,3 +387,187 @@ def test_side_staking(
...
+    spot_prices = []
+
+    def add_spot_price():
+        spot_prices.append(
+            bpool.get_spot_price(
+                ocean_token.address, erc20_token.address, swap_market_fee
+            )
+        )
+
+    add_spot_price()
```
Comment (DMats) [inline ocean_lib/models/test/test_side_staking.py:461]: Can you add some comments to each of these steps explaining what the test is trying to do?

File: tests/resources/helper_functions.py [hunk 9]
```diff
@@ -446,3 +447,113 @@ def create_nft_erc20_with_pool(
...
+def join_pool_with_max_base_token(bpool, web3, base_token, wallet, amt: int = 0):
+    """
+    Join pool with max base token if amt is 0. Otherwise join pool with amt base token.
+    """
+    pool_token_out_balance = bpool.get_balance(
+        base_token.address
+    )  # pool base token balance
+    max_out_ratio = bpool.get_max_out_ratio()  # max ratio
+
+    max_out_ratio_limit = int(from_wei(max_out_ratio, "ether") * pool_token_out_balance)
```
Comment (DMats) [inline tests/resources/helper_functions.py:461]: Instead of using `eth_utils.from_wei` from , can you instead use `ocean_lib.web3_internal.currency.from_wei`
https://github.com/oceanprotocol/ocean.py/blob/v4main/ocean_lib/web3_internal/currency.py#L40-L46

This eliminates the need for the "ether" argument and leads to more consistency throughout the ocean.py tests.

**Review by DMats — CHANGES_REQUESTED** [review]
Just a few minor comments, but other than that looks really good!

**Review by DMats — APPROVED** [review]
